server/src/main/java/org/elasticsearch/action/ingest/SimulateExecutionService.java

```diff
@@ -41,40 +43,44 @@
+            List<SimulateProcessorResult> processorResultList = new CopyOnWriteArrayList<>();
             CompoundProcessor verbosePipelineProcessor = decorate(pipeline.getCompoundProcessor(), processorResultList);
-            try {
-                Pipeline verbosePipeline = new Pipeline(pipeline.getId(), pipeline.getDescription(), pipeline.getVersion(),
-                    verbosePipelineProcessor);
-                ingestDocument.executePipeline(verbosePipeline);
-                return new SimulateDocumentVerboseResult(processorResultList);
-            } catch (Exception e) {
-                return new SimulateDocumentVerboseResult(processorResultList);
-            }
+            Pipeline verbosePipeline = new Pipeline(pipeline.getId(), pipeline.getDescription(), pipeline.getVersion(),
+                verbosePipelineProcessor);
+            ingestDocument.executePipeline(verbosePipeline, (result, e) -> {
+                handler.accept(new SimulateDocumentVerboseResult(processorResultList), e);
+            });
         } else {
-            try {
-                pipeline.execute(ingestDocument);
-                return new SimulateDocumentBaseResult(ingestDocument);
-            } catch (Exception e) {
-                return new SimulateDocumentBaseResult(e);
-            }
+            pipeline.execute(ingestDocument, (result, e) -> {
+                if (e == null) {
+                    handler.accept(new SimulateDocumentBaseResult(ingestDocument), null);
+                } else {
+                    handler.accept(new SimulateDocumentBaseResult(e), null);
+                }
+            });
         }
     }
 
     public void execute(SimulatePipelineRequest.Parsed request, ActionListener<SimulatePipelineResponse> listener) {
-        threadPool.executor(THREAD_POOL_NAME).execute(new ActionRunnable<SimulatePipelineResponse>(listener) {
+        threadPool.executor(THREAD_POOL_NAME).execute(new ActionRunnable<>(listener) {
             @Override
-            protected void doRun() throws Exception {
-                List<SimulateDocumentResult> responses = new ArrayList<>();
+            protected void doRun() {
+                final AtomicInteger counter = new AtomicInteger();
+                final List<SimulateDocumentResult> responses = new CopyOnWriteArrayList<>();
                 for (IngestDocument ingestDocument : request.getDocuments()) {
-                    SimulateDocumentResult response = executeDocument(request.getPipeline(), ingestDocument, request.isVerbose());
-                    if (response != null) {
-                        responses.add(response);
-                    }
+                    executeDocument(request.getPipeline(), ingestDocument, request.isVerbose(), (response, e) -> {
+                        if (response != null) {
+                            responses.add(response);
+                        }
+                        if (counter.incrementAndGet() == request.getDocuments().size()) {
```

Comment:
Will this still work when we fix https://github.com/elastic/elasticsearch/issues/36150 ?

Reply:
Yes, what we need to do the fix that issue is to use the `result` argument (and deal with the fact that `ingestDocument` can be null in the simulate response classes). That is null in case of document is dropped, now we always use the provided `ingestDocument`.